presto-main/src/main/java/com/facebook/presto/operator/scalar/sql/ArraySqlFunctions.java

```diff
@@ -118,4 +120,34 @@ public static String arrayHasDuplicatesBigint()
     {
         return "RETURN cardinality(array_duplicates(input)) > 0";
     }
-}
+
+    @SqlInvokedScalarFunction(value = "array_max_by", deterministic = true, calledOnNullInput = true)
+    @Description("Get the maximum value of array, by using a specific transformation function")
+    @TypeParameter("T")
+    @TypeParameter("U")
+    @SqlParameters({
+            @SqlParameter(name = "input", type="array(T)"),
+            @SqlParameter(name = "f", type="function(T, U)"),
+            }
+    )
+    @SqlType("T")
+    public static String arrayMaxBy()
+    {
+        return "RETURN input[ ARRAY_MAX(FILTER(ZIP(TRANSFORM(input, f), SEQUENCE(1, CARDINALITY(input))), x -> x[1] IS NOT NULL))[2] ]";
+    }
+
+    @SqlInvokedScalarFunction(value = "array_min_by", deterministic = true, calledOnNullInput = true)
+    @Description("Get the minimum value of array, by using a specific transformation function")
+    @TypeParameter("T")
+    @TypeParameter("U")
+    @SqlParameters({
+            @SqlParameter(name = "input", type="array(T)"),
+            @SqlParameter(name = "f", type="function(T, U)"),
+    }
+    )
+    @SqlType("T")
+    public static String arrayMinBy()
+    {
+        return "RETURN input[ ARRAY_MIN(FILTER(ZIP(TRANSFORM(input, f), SEQUENCE(1, CARDINALITY(input))), x -> x[1] IS NOT NULL))[2] ]";
```

Comment:
How about:

ARRAY_MIN(TRANSFORM(input, x->ROW(f(x), x)))[2]

Reply:
Or did you say that's what's not working?

Reply:
Yes can't do f(x) here

Reply:
ARRAY_MIN(zip(TRANSFORM(input, f), input))[2]

Why is null special?

Reply:
> ARRAY_MIN(zip(TRANSFORM(input, f), input))[2]

This works for simple cases but breaks when `input` itself is not orderable (T is not necessarily orderable so row(U, T) is not orderable). That's why I zipped with the index instead

> Why is null special?

This is to cover cases where f(t) is null, for example something like
`array_min_by(array[1,2,3,null], i -> i)`
then the pairs (f(t), i) can't be ordered since it fails with
```
com.facebook.presto.spi.PrestoException: null value at position 0
	at com.facebook.presto.type.TypeUtils.checkElementNotNull(TypeUtils.java:177)
	at com.facebook.presto.operator.scalar.RowComparisonOperator.compare(RowComparisonOperator.java:65)
	at com.facebook.presto.operator.scalar.RowLessThanOperator.less(RowLessThanOperator.java:59)
```

Reply:
Hmm - I'm in two minds here - I think we should fail here. NULL is just not ok for comparisons.

Reply:
Hm, so `array_min`/`array_max` currently return NULL if there array has a null value. We can probably do the same thing here? `IF(any_match(transform(input, f), x -> x IS NULL), NULL, <current logic>)`
However `min` and `max` (aggregation functions) does filter out the nulls

Reply:
I would really like to avoid the second scan. How about:

RETURN input[ ARRAY_MIN(
             ZIP_WITH(TRANSFORM(input, f), 
                               SEQUENCE(1, CARDINALITY(input)), 
                               (x, y)->IF(x IS NULL, NULL, (x y)))]

Reply:
And yes we should match the scalar versions ARRAY_MAX and ARRAY_MIN semantics here, not the min/max aggregtations

Reply:
> I would really like to avoid the second scan. How about:
> 
> RETURN input[ ARRAY_MIN( ZIP_WITH(TRANSFORM(input, f), SEQUENCE(1, CARDINALITY(input)), (x, y)->IF(x IS NULL, NULL, (x, y)))]

Yeah that's more clever! I updated the code and unit tests to match that logic